docs/en/setup/EnvVars.md
# Supported Environment Variables

Below is the full list of supported environment variables you can set to 
customize the agent behavior, please read the descriptions for what they can achieve.

 | Environment Variable                                | Description                                                                                                                                                                                                                                                                                             | Default                                                             |
|:----------------------------------------------------|:--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:--------------------------------------------------------------------|
| `SW_AGENT_NAME`                                     | The name of the Python service                                                                                                                                                                                                                                                                          | `Python Service Name`                                               |
| `SW_AGENT_INSTANCE`                                 | The name of the Python service instance (instance name will be followed by a f'-{os.getpid()}'                                                                                                                                                                                                          | Randomly generated - pid                                            |
| `SW_AGENT_NAMESPACE`                                | The agent namespace of the Python service                                                                                                                                                                                                                                                               | unset                                                               |
| `SW_AGENT_COLLECTOR_BACKEND_SERVICES`               | The backend OAP server address                                                                                                                                                                                                                                                                          | `127.0.0.1:11800`                                                   |
| `SW_AGENT_FORCE_TLS`                                | Use TLS for communication with server (no cert required)                                                                                                                                                                                                                                                | `False`                                                             |
| `SW_AGENT_PROTOCOL`                                 | The protocol to communicate with the backend OAP, `http`, `grpc` or `kafka`, **we highly suggest using `grpc` in production as it's well optimized than `http`**. The `kafka` protocol provides an alternative way to submit data to the backend.                                                       | `grpc`                                                              |
| `SW_AGENT_AUTHENTICATION`                           | The authentication token to verify that the agent is trusted by the backend OAP, as for how to configure the backend, refer to [the yaml](https://github.com/apache/skywalking/blob/4f0f39ffccdc9b41049903cc540b8904f7c9728e/oap-server/server-bootstrap/src/main/resources/application.yml#L155-L158). | unset                                                               |
| `SW_AGENT_LOGGING_LEVEL`                            | The logging level, could be one of `CRITICAL`, `FATAL`, `ERROR`, `WARN`(`WARNING`), `INFO`, `DEBUG`                                                                                                                                                                                                     | `INFO`                                                              |
| `SW_AGENT_DISABLE_PLUGINS`                          | The name patterns in CSV pattern, plugins whose name matches one of the pattern won't be installed                                                                                                                                                                                                      | `''`                                                                |
| `SW_AGENT_MAX_BUFFER_SIZE`                          | The maximum queue backlog size for sending the segment data to backend, segments beyond this are silently dropped                                                                                                                                                                                       | `'10000'`                                                           |
| `SW_HTTP_IGNORE_METHOD`                             | Comma-delimited list of http methods to ignore (GET, POST, HEAD, OPTIONS, etc...)                                                                                                                                                                                                                       | ``                                                                  |
| `SW_SQL_PARAMETERS_LENGTH`                          | The maximum length of the collected parameter, parameters longer than the specified length will be truncated, length 0 turns off parameter tracing                                                                                                                                                      | `0`                                                                 |
| `SW_PYMONGO_TRACE_PARAMETERS`                       | Indicates whether to collect the filters of pymongo                                                                                                                                                                                                                                                     | `False`                                                             |
| `SW_PYMONGO_PARAMETERS_MAX_LENGTH`                  | The maximum length of the collected filters, filters longer than the specified length will be truncated                                                                                                                                                                                                 | `512`                                                               |
| `SW_IGNORE_SUFFIX`                                  | If the operation name of the first span is included in this set, this segment should be ignored.                                                                                                                                                                                                        | `.jpg,.jpeg,.js,.css,.png,.bmp,.gif,.ico,.mp3,.mp4,.html,.svg`      |
| `SW_FLASK_COLLECT_HTTP_PARAMS`                      | This config item controls that whether the Flask plugin should collect the parameters of the request.                                                                                                                                                                                                   | `false`                                                             |
| `SW_DJANGO_COLLECT_HTTP_PARAMS`                     | This config item controls that whether the Django plugin should collect the parameters of the request.                                                                                                                                                                                                  | `false`                                                             |
| `SW_HTTP_PARAMS_LENGTH_THRESHOLD`                   | When `COLLECT_HTTP_PARAMS` is enabled, how many characters to keep and send to the OAP backend, use negative values to keep and send the complete parameters, NB. this config item is added for the sake of performance.                                                                                | `1024`                                                              |
| `SW_CORRELATION_ELEMENT_MAX_NUMBER`                 | Max element count of the correlation context.                                                                                                                                                                                                                                                           | `3`                                                                 |
| `SW_CORRELATION_VALUE_MAX_LENGTH`                   | Max value length of correlation context element.                                                                                                                                                                                                                                                        | `128`                                                               |
| `SW_TRACE_IGNORE`                                   | This config item controls that whether the trace should be ignore                                                                                                                                                                                                                                       | `false`                                                             |
| `SW_TRACE_IGNORE_PATH`                              | You can setup multiple URL path patterns, The endpoints match these patterns wouldn't be traced. the current matching rules follow Ant Path match style , like /path/*, /path/**, /path/?.                                                                                                              | `''`                                                                |
| `SW_ELASTICSEARCH_TRACE_DSL`                        | If true, trace all the DSL(Domain Specific Language) in ElasticSearch access, default is false                                                                                                                                                                                                          | `false`                                                             |
| `SW_KAFKA_REPORTER_BOOTSTRAP_SERVERS`               | A list of host/port pairs to use for establishing the initial connection to the Kafka cluster. It is in the form host1:port1,host2:port2,...                                                                                                                                                            | `localhost:9092`                                                    |
| `SW_KAFKA_REPORTER_TOPIC_MANAGEMENT`                | Specifying Kafka topic name for service instance reporting and registering.                                                                                                                                                                                                                             | `skywalking-managements`                                            |
| `SW_KAFKA_REPORTER_TOPIC_SEGMENT`                   | Specifying Kafka topic name for Tracing data.                                                                                                                                                                                                                                                           | `skywalking-segments`                                               |
| `SW_KAFKA_REPORTER_TOPIC_LOG`                       | Specifying Kafka topic name for Log data.                                                                                                                                                                                                                                                               | `skywalking-logs`                                                   |
| `SW_KAFKA_REPORTER_CONFIG_key`                      | The configs to init KafkaProducer. it support the basic arguments (whose type is either `str`, `bool`, or `int`) listed [here](https://kafka-python.readthedocs.io/en/master/apidoc/KafkaProducer.html#kafka.KafkaProducer)                                                                             | unset                                                               |
| `SW_CELERY_PARAMETERS_LENGTH`                       | The maximum length of `celery` functions parameters, longer than this will be truncated, 0 turns off                                                                                                                                                                                                    | `512`                                                               |
| `SW_AGENT_PROFILE_ACTIVE`                           | If `True`, Python agent will enable profile when user create a new profile task. Otherwise disable profile.                                                                                                                                                                                             | `True`                                                              |
| `SW_PROFILE_TASK_QUERY_INTERVAL`                    | The number of seconds between two profile task query.                                                                                                                                                                                                                                                   | `20`                                                                |
| `SW_AGENT_PROFILE_MAX_PARALLEL`                     | The number of parallel monitor segment count.                                                                                                                                                                                                                                                           | `5`                                                                 |
| `SW_AGENT_PROFILE_DURATION`                         | The maximum monitor segment time(minutes), if current segment monitor time out of limit, then stop it.                                                                                                                                                                                                  | `10`                                                                |
| `SW_AGENT_PROFILE_DUMP_MAX_STACK_DEPTH`             | The number of max dump thread stack depth                                                                                                                                                                                                                                                               | `500`                                                               |
| `SW_AGENT_PROFILE_SNAPSHOT_TRANSPORT_BUFFER_SIZE`   | The number of snapshot transport to backend buffer size                                                                                                                                                                                                                                                 | `50`                                                                |
| `SW_AGENT_LOG_REPORTER_ACTIVE`                      | If `True`, Python agent will report collected logs to the OAP or Satellite. Otherwise, it disables the feature.                                                                                                                                                                                         | `False`                                                             |
| `SW_AGENT_LOG_REPORTER_SAFE_MODE`                   | If `True`, Python agent will filter out HTTP basic auth information from log records. Otherwise, it disables the feature due to potential performance impact brought by regular expression                                                                                                              | `False`                                                             |
| `SW_AGENT_LOG_REPORTER_BUFFER_SIZE`                 | The maximum queue backlog size for sending log data to backend, logs beyond this are silently dropped.                                                                                                                                                                                                  | `10000`                                                             |
| `SW_AGENT_LOG_REPORTER_LEVEL`                       | This config specifies the logger levels of concern, any logs with a level below the config will be ignored.                                                                                                                                                                                             | `WARNING`                                                           |
| `SW_AGENT_LOG_REPORTER_IGNORE_FILTER`               | This config customizes whether to ignore the application-defined logger filters, if `True`, all logs are reported disregarding any filter rules.                                                                                                                                                        | `False`                                                             |
| `SW_AGENT_LOG_REPORTER_FORMATTED`                   | If `True`, the log reporter will transmit the logs as formatted. Otherwise, puts logRecord.msg and logRecord.args into message content and tags(`argument.n`), respectively. Along with an `exception` tag if an exception was raised.                                                                  | `True`                                                              |
| `SW_AGENT_LOG_REPORTER_LAYOUT`                      | The log reporter formats the logRecord message based on the layout given.                                                                                                                                                                                                                               | `%(asctime)s [%(threadName)s] %(levelname)s %(name)s - %(message)s` |
| `SW_AGENT_CAUSE_EXCEPTION_DEPTH`                    | This config limits agent to report up to `limit` stacktrace, please refer to [Python traceback](https://docs.python.org/3/library/traceback.html#traceback.print_tb) for more explanations.                                                                                                             | `10`                                                                | 
| `SW_PYTHON_BOOTSTRAP_PROPAGATE`                     | This config controls the child process agent bootstrap behavior in `sw-python` CLI, if set to `False`, a valid child process will not boot up a SkyWalking Agent. Please refer to the [CLI Guide](CLI.md) for details.                                                                                  | unset                                                               |
| `SW_FASTAPI_COLLECT_HTTP_PARAMS`                    | This config item controls that whether the FastAPI plugin should collect the parameters of the request.                                                                                                                                                                                                 | `false`                                                             |
| `SW_BOTTLE_COLLECT_HTTP_PARAMS`                    | This config item controls that whether the Bottle plugin should collect the parameters of the request.                                                                                                                                                                                                  | `false`                                                             |


CHANGELOG.md
## Change Logs

### 1.0.0
- Feature:
  - Drop support for Python 3.6
  - Add support for Python 3.11 (Pending)
  - Add MeterReportService (gRPC, Kafka reporter) (default:disabled) (#231, #236, #241, #243)
  - Add reporter for PVM runtime metrics (default:disabled, gRPC and Kafka only, HTTP is not supported yet) (#238, #247)
  - Add Greenlet profiler (#246)
  - Add test and support for Python Slim base images (#249)
  - Sw-python CLI now automatically handles Gunicorn and uWSGI pre-forking behavior. (#256)

- Plugins:
  - Add aioredis, aiormq, amqp, asyncpg, aio-pika, kombu RMQ plugins (#230 Missing test coverage) 
  - Add Confluent Kafka plugin (#233 Missing test coverage) 

- Fixes:
  - Allow RabbitMQ BlockingChannel.basic_consume() to link with outgoing spans (#224)
  - Fix RabbitMQ basic_get bug (#225, #226)
  - Fix case when tornado socket name is None (#227)
  - Fix misspelled text "PostgreSLQ" -> "PostgreSQL" in Postgres-related plugins (#234)
  - Make sure `span.component` initialized as Unknown rather than 0 (#242)
  - Ignore websocket connections inside fastapi temporarily (#244, issue#9724)
  - Fix Kafka-python plugin SkyWalking self reporter ignore condition (#249)

- Docs:
  - New documentation on how to test locally (#222)
  - New documentation on the newly added meter reporter feature (#240)
  - New documentation on the newly added greenlet profiler and the original threading profiler (#250)
  - Overhaul documentation on development setup and testing (#249)
  - Overhaul documentation on how to use agent with preforking servers [Gunicorn/uWSGI] (#256)

- Others:
  - Pin CI SkyWalking License Eye (#221)
  - Fix dead link due to the 'next' url change (#235)
  - Pin CI SkyWalking Infra-E2E (#251)
  - Sync OAP, SWCTL versions in E2E and fix test cases (#249)
  - Overhaul development flow with Poetry (#249)
  - Fix grpcio-tools generated message type (#253)

### 0.8.0
- Feature:
  - Update mySQL plugin to support two different parameter keys. (#186)
  - Add a `SW_AGENT_LOG_REPORTER_SAFE_MODE` option to control the HTTP basic auth credential filter (#200)

- Plugins:
  - Add Psycopg(3.x) support (#168)
  - Add MySQL support (#178)
  - Add FastAPI support (#181)
  - Drop support for flask 1.x due to dependency issue in Jinja2 and EOL (#195)
  - Add Bottle support (#214)

- Fixes:
  - Spans now correctly reference finished parents (#161)
  - Remove potential password leak from Aiohttp outgoing url (#175)
  - Handle error when REMOTE_PORT is missing in Flask (#176)
  - Fix sw-rabbitmq TypeError when there are no headers (#182)
  - Fix agent bootstrap traceback not shown in sw-python CLI (#183)
  - Fix local log stack depth overridden by agent log formatter (#192)
  - Fix typo that cause user sitecustomize.py not loaded (#193)
  - Fix instance property wrongly shown as UNKNOWN in OAP (#194)
  - Fix multiple components inconsistently named on SkyWalking UI (#199)
  - Fix SW_AGENT_LOGGING_LEVEL not properly set during startup (#196)
  - Unify the http tag name with other agents (#208)
  - Remove namespace to instance properties and add pid property (#205)
  - Fix the properties are not set correctly (#198)
  - Improved ignore path regex (#210)
  - Fix sw_psycopg2 register_type() (#211)
  - Fix psycopg2 register_type() second arg default (#212)
  - Enhance Traceback depth (#206)
  - Set spans whose http code > 400 to error (#187)

- Docs:
  - Add a FAQ doc on `how to use with uwsgi` (#188)

- Others:
  - Refactor current Python agent docs to serve on SkyWalking official website (#162)
  - Refactor SkyWalking Python to use the CLI for CI instead of legacy setup (#165)
  - Add support for Python 3.10 (#167)
  - Move flake configs all together (#169)
  - Introduce another set of flake8 extensions (#174)
  - Add E2E test coverage for trace and logging (#199)
  - Now Log reporter `cause_exception_depth` traceback limit defaults to 10
  - Enable faster CI by categorical parallelism (#170)

### 0.7.0

- Feature:
    - Support collecting and reporting logs to backend (#147)
    - Support profiling Python method level performance (#127
    - Add a new `sw-python` CLI that enables agent non-intrusive integration (#156)
    - Add exponential reconnection backoff strategy when OAP is down (#157)
    - Support ignoring traces by http method (#143)
    - `NoopSpan` on queue full, propagation downstream (#141)
    - Support agent namespace. (#126)
    - Support secure connection option for GRPC and HTTP (#134)

- Plugins:
    - Add Falcon Plugin (#146)
    - Update `sw_pymongo.py` to be compatible with cluster mode (#150)
    - Add Python celery plugin (#125)
    - Support tornado5+ and tornado6+ (#119)

- Fixes:
    - Remove HTTP basic auth credentials from log, stacktrace, segment (#152)
    - Fix `@trace` decorator not work (#136)
    - Fix grpc disconnect, add `SW_AGENT_MAX_BUFFER_SIZE` to control buffer queue size (#138)

- Others:
    - Chore: bump up `requests` version to avoid license issue (#142)
    - Fix module wrapt as normal install dependency (#123)
    - Explicit component inheritance (#132)
    - Provide dockerfile & images for easy integration in containerized scenarios (#159)

### 0.6.0

- Fixes:
    - Segment data loss when gRPC timing out. (#116)
    - `sw_tornado` plugin async handler status set correctly. (#115)
    - `sw_pymysql` error when connection haven't db. (#113)

### 0.5.0

- New plugins
    - Pyramid Plugin (#102)
    - AioHttp Plugin (#101)
    - Sanic Plugin (#91)

- API and enhancements
    - `@trace` decorator supports `async` functions
    - Supports async task context
    - Optimized path trace ignore
    - Moved exception check to `Span.__exit__`
    - Moved Method & Url tags before requests

- Fixes:
    - `BaseExceptions` not recorded as errors
    - Allow pending data to send before exit
    - `sw_flask` general exceptions handled
    - Make `skywalking` logging Non-global

- Chores and tests
    - Make tests really run on specified Python version

skywalking/config.py
#
# Licensed to the Apache Software Foundation (ASF) under one or more
# contributor license agreements.  See the NOTICE file distributed with
# this work for additional information regarding copyright ownership.
# The ASF licenses this file to You under the Apache License, Version 2.0
# (the "License"); you may not use this file except in compliance with
# the License.  You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
import os
import re
import uuid

from typing import List, Pattern

QUEUE_TIMEOUT: int = 1

RE_IGNORE_PATH: Pattern = re.compile('^$')
RE_HTTP_IGNORE_METHOD: Pattern = RE_IGNORE_PATH

options = None  # here to include 'options' in globals
options = globals().copy()  # THIS MUST PRECEDE DIRECTLY BEFORE LIST OF CONFIG OPTIONS!

# Core level configurations
service_name: str = os.getenv('SW_AGENT_NAME') or 'Python Service Name'
service_instance: str = f"{os.getenv('SW_AGENT_INSTANCE')}" or str(uuid.uuid1()).replace('-', '')

# fixme 加上fu'gaf
agent_namespace: str = os.getenv('SW_AGENT_NAMESPACE')
collector_address: str = os.getenv('SW_AGENT_COLLECTOR_BACKEND_SERVICES') or '127.0.0.1:11800'
kafka_bootstrap_servers: str = os.getenv('SW_KAFKA_REPORTER_BOOTSTRAP_SERVERS') or 'localhost:9092'
kafka_topic_management: str = os.getenv('SW_KAFKA_REPORTER_TOPIC_MANAGEMENT') or 'skywalking-managements'
kafka_topic_segment: str = os.getenv('SW_KAFKA_REPORTER_TOPIC_SEGMENT') or 'skywalking-segments'
kafka_topic_log: str = os.getenv('SW_KAFKA_REPORTER_TOPIC_LOG') or 'skywalking-logs'
kafka_topic_meter: str = os.getenv('SW_KAFKA_REPORTER_TOPIC_METER') or 'skywalking-meters'
force_tls: bool = os.getenv('SW_AGENT_FORCE_TLS', '').lower() == 'true'
protocol: str = (os.getenv('SW_AGENT_PROTOCOL') or 'grpc').lower()
authentication: str = os.getenv('SW_AGENT_AUTHENTICATION')
logging_level: str = os.getenv('SW_AGENT_LOGGING_LEVEL') or 'INFO'
disable_plugins: List[str] = (os.getenv('SW_AGENT_DISABLE_PLUGINS') or '').split(',')
max_buffer_size: int = int(os.getenv('SW_AGENT_MAX_BUFFER_SIZE', '10000'))
trace_ignore_path: str = os.getenv('SW_TRACE_IGNORE_PATH') or ''
ignore_suffix: str = os.getenv('SW_IGNORE_SUFFIX') or '.jpg,.jpeg,.js,.css,.png,.bmp,.gif,.ico,.mp3,' \
                                                      '.mp4,.html,.svg '
correlation_element_max_number: int = int(os.getenv('SW_CORRELATION_ELEMENT_MAX_NUMBER') or '3')
correlation_value_max_length: int = int(os.getenv('SW_CORRELATION_VALUE_MAX_LENGTH') or '128')

# Plugin configurations
sql_parameters_length: int = int(os.getenv('SW_SQL_PARAMETERS_LENGTH') or '0')
pymongo_trace_parameters: bool = os.getenv('SW_PYMONGO_TRACE_PARAMETERS') == 'True'
pymongo_parameters_max_length: int = int(os.getenv('SW_PYMONGO_PARAMETERS_MAX_LENGTH') or '512')
elasticsearch_trace_dsl: bool = os.getenv('SW_ELASTICSEARCH_TRACE_DSL') == 'True'

http_params_length_threshold: int = int(os.getenv('SW_HTTP_PARAMS_LENGTH_THRESHOLD') or '1024')
http_ignore_method: str = os.getenv('SW_HTTP_IGNORE_METHOD', '').upper()
flask_collect_http_params: bool = os.getenv('SW_FLASK_COLLECT_HTTP_PARAMS') == 'True'
sanic_collect_http_params: bool = os.getenv('SW_SANIC_COLLECT_HTTP_PARAMS') == 'True'
django_collect_http_params: bool = os.getenv('SW_DJANGO_COLLECT_HTTP_PARAMS') == 'True'
fastapi_collect_http_params: bool = os.getenv('SW_FASTAPI_COLLECT_HTTP_PARAMS') == 'True'
bottle_collect_http_params: bool = os.getenv('SW_BOTTLE_COLLECT_HTTP_PARAMS') == 'True'

celery_parameters_length: int = int(os.getenv('SW_CELERY_PARAMETERS_LENGTH') or '512')

# profiling configurations
get_profile_task_interval: int = int(os.getenv('SW_PROFILE_TASK_QUERY_INTERVAL') or '20')
profile_active: bool = os.getenv('SW_AGENT_PROFILE_ACTIVE') != 'False'
profile_max_parallel: int = int(os.getenv('SW_AGENT_PROFILE_MAX_PARALLEL') or '5')
profile_duration: int = int(os.getenv('SW_AGENT_PROFILE_DURATION') or '10')
profile_dump_max_stack_depth: int = int(os.getenv('SW_AGENT_PROFILE_DUMP_MAX_STACK_DEPTH') or '500')
profile_snapshot_transport_buffer_size: int = int(os.getenv('SW_AGENT_PROFILE_SNAPSHOT_TRANSPORT_BUFFER_SIZE') or '50')

# log reporter configurations
log_reporter_active: bool = os.getenv('SW_AGENT_LOG_REPORTER_ACTIVE') == 'True'
log_reporter_safe_mode: bool = os.getenv('SW_AGENT_LOG_REPORTER_SAFE_MODE') == 'True'
log_reporter_max_buffer_size: int = int(os.getenv('SW_AGENT_LOG_REPORTER_BUFFER_SIZE') or '10000')
log_reporter_level: str = os.getenv('SW_AGENT_LOG_REPORTER_LEVEL') or 'WARNING'
log_reporter_ignore_filter: bool = os.getenv('SW_AGENT_LOG_REPORTER_IGNORE_FILTER') == 'True'
log_reporter_formatted: bool = os.getenv('SW_AGENT_LOG_REPORTER_FORMATTED') != 'False'
log_reporter_layout: str = os.getenv('SW_AGENT_LOG_REPORTER_LAYOUT') or \
                           '%(asctime)s [%(threadName)s] %(levelname)s %(name)s - %(message)s'
# This configuration is shared by log reporter and tracer
cause_exception_depth: int = int(os.getenv('SW_AGENT_CAUSE_EXCEPTION_DEPTH') or '10')

# meter reporter configurations
meter_reporter_active: bool = os.getenv('SW_AGENT_METER_REPORTER_ACTIVE') == 'True'
meter_reporter_max_buffer_size: int = int(os.getenv('SW_AGENT_METER_REPORTER_BUFFER_SIZE') or '10000')
meter_reporter_peroid: int = int(os.getenv('SW_AGENT_METER_REPORTER_PEROID') or '20')
pvm_meter_reporter_active: bool = os.getenv('SW_AGENT_PVM_METER_REPORTER_ACTIVE') == 'True'

options = {key for key in globals() if key not in options}  # THIS MUST FOLLOW DIRECTLY AFTER LIST OF CONFIG OPTIONS!


def init(**kwargs):
    glob = globals()

    for key, val in kwargs.items():
        if key not in options:
            raise KeyError(f'invalid config option {key}')

        glob[key] = val


def finalize():
    reesc = re.compile(r'([.*+?^=!:${}()|\[\]\\])')
    suffix = r'^.+(?:' + '|'.join(reesc.sub(r'\\\1', s.strip()) for s in ignore_suffix.split(',')) + ')$'
    method = r'^' + '|'.join(s.strip() for s in http_ignore_method.split(',')) + '$'
    path = '^(?:' + \
           '|'.join(  # replaces ","
               '/(?:[^/]*/)*'.join(  # replaces "/**/"
                   '(?:(?:[^/]+/)*[^/]+)?'.join(  # replaces "**"
                       '[^/]*'.join(  # replaces "*"
                           '[^/]'.join(  # replaces "?"
                               reesc.sub(r'\\\1', s) for s in p3.split('?')
                           ) for p3 in p2.split('*')
                       ) for p2 in p1.strip().split('**')
                   ) for p1 in p0.split('/**/')
               ) for p0 in trace_ignore_path.split(',')
           ) + ')$'

    global RE_IGNORE_PATH, RE_HTTP_IGNORE_METHOD
    RE_IGNORE_PATH = re.compile(f'{suffix}|{path}')
    RE_HTTP_IGNORE_METHOD = re.compile(method, re.IGNORECASE)


def ignore_http_method_check(method: str):
    return RE_HTTP_IGNORE_METHOD.match(method)
